util/svdgen.py

```diff
@@ -0,0 +1,604 @@
+            create('registers', *registers(ip['registers'])),
+            ET.Comment('end of %s' % module['name']),
+            name        = module['name'],
+            baseAddress = module['base_addr'])
+
+def cpu() -> ET.Element:
+    """Generate a <cpu> element. Currently all values are hardcoded as
+    they are not included in either the top level or IP module HJSON."""
+
+    # TODO: allow customizing these.
+
+    # RISCV is not present in pysvd so this would fail validation. We
+    # set it to the (inaccurate) value `other`, and patch it up after
+    # performing the validation step.
+    name = pysvd.type.cpuName.other
+
+    return create('cpu',
+            name                = name,
+            endian              = pysvd.type.endian.little,
+            revision            = 0,
+            mpuPresent          = "true",
+            fpuPresent          = "false",
+            nvicPrioBits        = 0,
+            vendorSystickConfig = "false",
+    )
+
+def device(top: hjson, ips: {'name': hjson}, version: str) -> ET.Element:
+    """Generate an SVD <device> node from the top-level HJSON. For each
+    module defined in `top` the corresponding peripheral definition is
+    generated by the IP block in `ips`.
+
+    SVD files need some basic information that isn't provided within the
+    HJSON descriptions. While they might be available elsewhere in the
+    configuration, it's simplest just to declare them here."""
+
+    # Cross reference module with IP to generate a peripheral
+    def p(m: hjson) -> ET.Element:
+        return peripheral(m, ips[m['type']])
+
+    return create(ET.Element('device', attrib = {
+                'schemaVersion': '1.1',
+                'xmlns:xs': 'http://www.w3.org/2001/XMLSchema-instance',
+                'xs:noNamespaceSchemaLocation': 'CMSIS-SVD.xsd',
+            }),
+            cpu(),
+            create('peripherals', *map(p, top['module'])),
+            vendor          = 'lowRISC OpenTitan',
+            vendorID        = 'lowRISC',
+            name            = top['name'],
+            version         = version,
+            description     = read_relative_path('..', 'README.md'),
+            width           = top['datawidth'],
+            size            = top['datawidth'],
+            addressUnitBits = 8)
+
+def generate(top: hjson, ips: {'name': hjson}, version: str, validate=True) -> ET.Element:
+    """Convert a top-level configuration and IP register definition set
+    into a System View Description ("SVD") file. SVD is an XML format,
+    originally defined by ARM to "formalize the description of the system
+    contained in Arm Cortex-M processor-based microcontrollers".¹ It has
+    since been extended for use on other platforms, and is used in other
+    fields such as Rust embedded programming.²
+
+    SVD is an XML-based format with standardized definitions for memory
+    layouts, register definitions, and similar microcontroller fields.
```

Comment:
I'd drop the background on SVD.  Keep a link to the CMSIS-SVD docs, but don't include direct quotes.